addons/website_event/models/event_event.py

```diff
@@ -116,28 +143,61 @@ def _compute_is_participating(self):
             registration;
         """
         current_visitor = self.env['website.visitor']._get_visitor_from_request()
-        base_domain = [('event_id', 'in', self.ids), ('state', 'in', ['open', 'done'])]
+        base_domain = [('state', 'in', ['open', 'done'])]
+        if self:
+            base_domain = expression.AND([[('event_id', 'in', self.ids)], base_domain])
         if self.env.user._is_public() and not current_visitor:
-            events = self.env['event.event']
-        elif self.env.user._is_public():
-            events = self.env['event.registration'].sudo().search(
-                expression.AND([base_domain, [('visitor_id', '=', current_visitor.id)]])
-            ).event_id
+            return self.env['event.event']
+        if current_visitor:
+            visitor_domain = [
+                '|',
+                ('partner_id', '=', self.env.user.partner_id.id),
+                ('visitor_id', '=', current_visitor.id),
+            ]
         else:
-            if current_visitor:
-                domain = [
-                    '|',
-                    ('partner_id', '=', self.env.user.partner_id.id),
-                    ('visitor_id', '=', current_visitor.id)
-                ]
+            visitor_domain = [('partner_id', '=', self.env.user.partner_id.id)]
+        is_participating_domain = expression.AND([visitor_domain, base_domain])
+
+        registrations_events = self.env['event.registration'].sudo()._read_group(is_participating_domain, ['event_id'])
+        return self.env['event.event'].concat(*[event[0] for event in registrations_events])
+
+    @api.depends_context('uid')
+    @api.depends('website_visibility')
+    def _compute_is_visible_on_website(self):
+        if all(event.website_visibility == 'public' for event in self):
+            self.is_visible_on_website = True
+            return
+        events_user_participating = self._fetch_is_participating_events()
+        for event in self:
+            if event.website_visibility == 'public' or event in events_user_participating:
```

Comment:
Actually we should be able to use "event.is_participating" here right? Would simplify a bit and the ORM should batch that nicely.